atlasdb-client/src/main/java/com/palantir/atlasdb/cache/DefaultOffHeapCache.java

```diff
@@ -0,0 +1,178 @@
+    }
+
+    @Override
+    public Optional<V> getIfPresent(K key) {
+        Optional<V> value = getValue(key);
+
+        if (value.isPresent()) {
+            taggedMetricRegistry.meter(CACHE_HIT).mark();
+        } else {
+            taggedMetricRegistry.meter(CACHE_MISS).mark();
+        }
+        return value;
+    }
+
+    private Optional<V> getValue(K key) {
+        V inFlight = inFlightRequests.get(key);
+        if (inFlight != null) {
+            return Optional.of(inFlight);
+        }
+
+        return persistentStore.get(cacheDescriptor.get().storeNamespace(), key);
+    }
+
+    private static <K, V> CacheDescriptor createNamespaceAndConstructCacheProposal(
+            LogicalPersistentStore<K, V> store) {
+        StoreNamespace proposal = store.createNamespace(OFF_CACHE_NAMESPACE);
+        return ImmutableCacheDescriptor.builder()
+                .currentSize(new AtomicInteger())
+                .storeNamespace(proposal)
+                .build();
+    }
+
+    private static MetricName constructCacheMetricName(String metricSuffix) {
+        return MetricName.builder()
+                .safeName(MetricRegistry.name(DefaultOffHeapCache.class, metricSuffix))
+                .build();
+    }
+
+    private static class WriteBatcher <K, V>
+            implements CoalescingRequestFunction<Map.Entry<K, V>, Map.Entry<K, V>> {
+        DefaultOffHeapCache<K, V> offHeapCache;
+
+        WriteBatcher(DefaultOffHeapCache<K, V> offHeapCache) {
+            this.offHeapCache = offHeapCache;
+        }
+
+        @Override
+        public Map<Map.Entry<K, V>, Map.Entry<K, V>> apply(Set<Map.Entry<K, V>> request) {
+            CacheDescriptor cacheDescriptor = offHeapCache.cacheDescriptor.get();
+            if (cacheDescriptor.currentSize().get() >= offHeapCache.maxSize.getAsLong()) {
+                offHeapCache.taggedMetricRegistry.counter(CACHE_NUKE).inc();
+                offHeapCache.clear();
+            }
+            cacheDescriptor = offHeapCache.cacheDescriptor.get();
+            try {
+                Map<K, V> response = offHeapCache.persistentStore.get(
+                        cacheDescriptor.storeNamespace(),
+                        request.stream().map(Map.Entry::getKey).collect(Collectors.toList()));
+
+                int sizeIncrease = Sets.difference(request, response.entrySet()).size();
+                offHeapCache.persistentStore.put(
+                        cacheDescriptor.storeNamespace(),
+                        ImmutableMap.copyOf(request));
+
+                cacheDescriptor.currentSize().addAndGet(sizeIncrease);
+            } catch (SafeIllegalArgumentException exception) {
+                // happens when a store is dropped by a concurrent call to clear
+                log.warn("Clear called concurrently, writing failed", exception);
+            } finally {
+                offHeapCache.inFlightRequests.clear();
```

Comment:
is this correct?

if you have a request that comes in for key K, whilst the autobatcher is in here, it'll be in the next batch. If you clear, then another request for key K can come in (still in the same batch, since we haven't returned) because it's not an in flight request anymore.

if it doesn't matter, then the in flight request map doesn't matter because autobatcher will do the right thing (you're using the coalescing one).

if it does matter, then you're gonna have to fix this

Reply:
I removed the `inflight`. As discussed this logic can be implemented for the specific timestamp cache.